### ISSUE TITLE
Idea: require reviews from selected people/teams if specific files/folders are touched by PR

### ISSUE THREAD

**Issue by lekkas** [resin-io-modules/resin-procbots#139]
This came up after a discussion with @alexandrosm as a way to enforce stricter reviews for anything affecting a component's build/deployment process. For instance, versionbot could explicitly require review from @resin-io/devops if any of

```
- .travis.yml
- circle.yml
- build/
```

has been changed by a PR. This would mitigate the risk of rogue scripts that extract sensitive environment variables (e.g. npm publishing keys)

**Comment by hedss** [comment]
Augment the current `73-27-review-configs` branch to include the following:

```
review_files:
  - files: <array | pattern>
    reviewers: [array of users]
```
Any specific reviewers either adds to (or creates) the `reviewers` list, so they are included on review for any PR that includes changes to those files.

**Comment by lurch** [comment]
But would this prevent the PR from automatically being built and run on the CI services in the first place?

**Comment by hedss** [comment]
@lurch @lekkas So I'd assumed this would just stop it from being merged to `master`, but actually after re-reading the original comment, it does **sound** as if we'd actually be trying to block and PR CI build occuring. Is that right, Kostas?

If so, I'd need to think about this, as I'm not sure we could do so easily. What we'd have to do is something like:
* Keep `master` branch protection as it is now, only merge if other status checks pass
* Ensure VersionBot is the sole status condition as branch protection for the *specific* branch the PR is on
* When reviewers have approved so that the status is successful, VB would then add the other CI systems as required *for the same branch*, which would then carry out builds and report back
* Rest of the process continues as required

I have a bit of a worry, which is by default VB assumes things fail unless otherwise proven (paranoia is goooood!). With this system, VB would be the only thing stopping builds occurring until approvals occur, when the other CI components would be added. On the one hand, no different to how it works now as `master` protection would still be active, on the other hand, VB has to optimistically work forwards when reviews occur.

**Comment by lurch** [comment]
Hmmm, it's something I've never set up myself (so I could be wrong) but I thought that when e.g. adding Travis or Appveyor to a repo, those CI services then automatically build _any_ PRs issued against that repo; and the branch-protection stuff only stops PRs being merged when a status-check fails, it doesn't determine which CI services actually get run against a PR?

Although maybe there's something in the Travis or Appveyor `.yml` files that can be used to tweak this always-build-every-PR behaviour?

And just to clarify: The protected env-vars used in [Travis](https://docs.travis-ci.com/user/environment-variables/#Defining-Variables-in-Repository-Settings) and [Appveyor](https://www.appveyor.com/docs/build-configuration/#secure-variables) builds are already protected to the extent that they're not available to PRs from external forks. AIUI this issue is concerned with trying to stop somebody with write-access on a repo (i.e. a resin.io employee) from submitting a PR that could leak a protected env-var.
Hmmm, interestingly (according to the documentation links above) it seems Travis's secure env-vars are only blocked in PRs from external repos (forks), whereas Appveyor's secure env-vars are blocked in all PR-builds (and are therefore presumably only available in merge-builds). In which case the protection-mechanism described in this issue would be sufficient for Appveyor but not for Travis? (AIUI)

P.S. I also found https://github.com/ChALkeR/notes/blob/master/Stealing-Travis-secure-variables.md which might be interesting / relevant ?

Pinging in @petrosagg too, as this is security-related.

**Comment by hedss** [comment]
@lurch Yes, exactly right on both the CI and protection, which is why if we're worried about keys etc., then we need to stop these CI systems building when the PR is opened. The main idea here is only the minimum number of people who need access to keys have them, so we're trying to prevent anyone with read/write from accessing the containers as they build. I believe this stems from a conversation Alex, Petros and Kostas had.

You're correct in the assumptions for Travis (and, more or less, the same is true of Circle).

**Comment by lurch** [comment]
Indeed, only giving people the minimum privileges needed (and no more) is always a worthy security goal.

**Comment by hedss** [comment]
@lekkas @alexandrosm I am now satisfied that the [`CODEOWNERS`](https://help.github.com/articles/about-codeowners/) file along with enabling code protected review status in a repository does exactly this. I've opened a new issue here https://github.com/resin-io/process/issues/105
